src/Geometry/CSG/CSG2.jl

```diff
@@ -0,0 +1,83 @@
+# MIT license
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# See LICENSE in the project root for full license information.
+
+import Base: ∩, ∪, -
+using LinearAlgebra
+
+export CSGTree2
+
+# TODO CSGComplement, CSGDifference
+abstract type CSGType end
+abstract type CSGLeaf <: CSGType end
+abstract type CSGUnion <: CSGType end
+abstract type CSGIntersection <: CSGType end
+
+"""
+    CSGData{T<:Real}
+
+A flexible struct that can hold the data for any kind of CSG node. Leaf nodes define both a `transform` and `surface`;
+non-leaf nodes leave the latter undefined. The inverse transform is calculated and stored in the `inverse` field.
+"""
+struct CSGData{T<:Real}
+    transform::Transform{T}
+    inverse::Transform{T}
+    surface::ParametricSurface{T,3}
+
+    # non-leaf node data
+    CSGData(transform::Transform{T}) where T = new{T}(transform, inv(transform))
+    CSGData{T}() where T = CSGData(identitytransform(T))
+
+    # leaf node data
+    function CSGData(surface::ParametricSurface{T,3}, transform::Transform{T} = identitytransform(T)) where T
+        return new{T}(transform, inv(transform), surface)
+    end
+end
+function Base.show(io::IO, data::CSGData)
+    print(io, "transform = $(data.transform ≈ I ? "I" : data.transform)")
+    isdefined(data, :surface) && print(io, ", surface = $(data.surface)")
+end
+
+"""
+    CSGTree{N<:CSGType,T<:Real}
+
+A binary tree representing a CSG structure. `N` defines the root node's type: Leaf, Union, Intersection or Difference.
+
+The leaf nodes of a `CSGTree` contain parametric surfaces. Non-leaf nodes define CSG operations between their children.
+At each node, it is possible to store a transform that propagates down the tree at evaluation.
+"""
+mutable struct CSGTree2{N<:CSGType,T<:Real} <: Primitive{T}
+    data::CSGData{T}
```

Comment:
These are all abstract/unsized types so we're going to have a real bad time with allocations as it is.
I think you'll need both left and right type parameterized, also the CSGData wrapper will probably need the surface type parameterized.